# Tarea - Programación lineal

### Diseño de la Dieta Óptima

Se quiere producir comida para gatos de la manera más barata, no obstante se debe también asegurar que se cumplan los datos requeridos de analisis nutricional. Por lo que se quiere variar la cantidad de cada ingrediente para cumplir con los estandares nutricionales. Los requisitos que se tienen es que en $100$ gramos, se deben tener **por lo menos** $8$ gramos de proteína y $6$ gramos de grasa. Así mismo, no se debe tener más de $2$ gramos de fibra y $0.4$ gramos de sal.  

Los datos nutricionales se pueden obtener de la siguiente tabla:

Ingrediente|Proteína|Grasa|Fibra|Sal
:----|----|----|----|----
Pollo|  10.0%|08.0%|00.1%|00.2%
Carne|  20.0%|10.0%|00.5%|00.5%
Cordero|15.0%|11.0%|00.5%|00.7%
Arroz|  00.0%|01.0%|10.0%|00.2%
Trigo|  04.0%|01.0%|15.0%|00.8%
Gel|    00.0%|00.0%|00.0%|00.0%

Los costos de cada producto son:

Ingrediente|Costo por gramo
:----|----
Pollo|$\$$0.013
Carne|$\$$0.008
Cordero|$\$$0.010
Arroz|$\$$0.002
Trigo|$\$$0.005
Gel|$\$$0.001    

Lo que se busca optimizar en este caso es la cantidad de productos que se debe utilizar en la comida de gato, minimizando el costo total. Para simplificar la notación use las siguientes variables: 

+ $x_1:$ Gramos de pollo  
+ $x_2:$ Gramos de carne  
+ $x_3:$ Gramos de cordero  
+ $x_4:$ Gramos de arroz  
+ $x_5:$ Gramos de trigo  
+ $x_6:$ Gramos de gel

La tarea consiste en plantear el problemade programación lineal que permita satisfacer las necesidades alimenticias del gato al tiempo que minimice el costo total y resolverlo con `linprog`.

In [1]:
import scipy.optimize as opt
import numpy as np

In [13]:
c=np.array([0.013,0.008,0.010,0.002,0.005,0.001])
A_ub=np.array([
            [ -0.1 ,  -0.2 ,  -0.15,  0   , -0.04,   0],
            [ -0.08,  -0.1 ,  -0.11, -0.01, -0.01,   0],
            [ 0.001, 0.005 , 0.005 ,  0.1 , 0.15 ,   0],
            [ 0.002, 0.005 , 0.007 , 0.002, 0.008,   0]
           ])
b_ub=np.array([-8,-6,2,0.4])

A_eq=np.array([[1,1,1,1,1,1]])
b_eq=np.array([100])

opt.linprog(c,A_ub,b_ub,A_eq,b_eq)

     con: array([3.01838554e-11])
     fun: 0.520000000021656
 message: 'Optimization terminated successfully.'
     nit: 8
   slack: array([4.00000000e+00, 7.70405961e-11, 1.70000000e+00, 1.00000000e-01])
  status: 0
 success: True
       x: array([4.50819666e-10, 6.00000000e+01, 4.08653402e-09, 2.29072688e-08,
       3.70730008e-10, 4.00000000e+01])

El precio mínimo de ***52 centavos*** por 100 g de producto se logra si se usan:
- 60 gramos de Carne
- 40 gramos de Gel

### Problema de distribución de energía eléctrica

La Comisión Federal de Electricidad **(CFE)** dispone de cuatro plantas de generación para satisfacer la demanda diaria eléctrica en cuatro ciudades, Guadalajara, León y Morelia. Las plantas $1$, $2$ y $3$ pueden satisfacer $80$, $40$ y $60$ millones de KW al día respectivamente. Las necesidades de las ciudades de Guadalajara, León y Morelia son de $70$, $40$ y $70$ millones de Kw al día respectivamente. 


Los costos asociados al envío de suministro energético por cada millón de Kw entre cada planta y cada ciudad son los registrados en la siguiente tabla. 

-|Guadalajara|León|Morelia
:----|----|----|----
Planta 1|5|2|7
Planta 2|3|6|6
Planta 3|6|1|2

Y por último, las restricciones del problema, van a estar dadas por las capacidades de oferta y demanda de cada planta (en millones de KW) y cada ciudad.

Para simplificar la notación use las siguientes variables:

+ $x_1$: Kw (en millones) distribuidos de la Planta 1 a Guadalajara
+ $x_2$: Kw (en millones) distribuidos de la Planta 1 a León
+ $x_3$: Kw (en millones) distribuidos de la Planta 1 a Morelia
+ $x_4$: Kw (en millones) distribuidos de la Planta 2 a Guadalajara
+ $x_5$: Kw (en millones) distribuidos de la Planta 2 a León
+ $x_6$: Kw (en millones) distribuidos de la Planta 2 a Morelia
+ $x_7$: Kw (en millones) distribuidos de la Planta 3 a Guadalajara
+ $x_8$: Kw (en millones) distribuidos de la Planta 3 a León
+ $x_9$: Kw (en millones) distribuidos de la Planta 3 a Morelia

La tarea consiste en plantear el problema de programación lineal que permita satisfacer las necesidades de todas las ciudades al tiempo que minimice los costos asociados a la distribución y resolverlo con `linprog`.

Lla ***función objetivo*** para minimizar es:
$$5x_1+2x_2+7x_3+3x_4+6x_5+6x_6+6x_7+x_8+2x_9$$

Las variables de la planta 1 son: $x_1,x_2,x_3$
y su ecuación es $x_1+x_2+x_3\leq80$

Las variables de la planta 2 son: $x_4,x_5,x_6$
y su ecuación es $x_4+x_5+x_6\leq40$

Las variables de la planta 1 son: $x_7,x_8,x_9$
y su ecuación es $x_7+x_8+x_9\leq60$

El consumo de Guadalajara está dado por $x_1+x_4+x_7\geq70$

El consumo de León está dado por $x_2+x_5+x_8\geq40$

El consumo de Morelia está dado por $x_3+x_6+x_9\geq70$

In [11]:
import scipy.optimize as opt
import numpy as np

c=np.array([5,2,7,3,6,6,6,1,2])

A=np.array([[ 1 , 1 , 1 , 0 , 0 , 0 , 0 , 0 , 0 ],
            [ 0 , 0 , 0 , 1 , 1 , 1 , 0 , 0 , 0 ],
            [ 0 , 0 , 0 , 0 , 0 , 0 , 1 , 1 , 1 ],
            [-1 , 0 , 0 ,-1 , 0 , 0 ,-1 , 0 , 0 ],
            [ 0 ,-1 , 0 , 0 ,-1 , 0 , 0 ,-1 , 0 ],
            [ 0 , 0 ,-1 , 0 , 0 ,-1 , 0 , 0 ,-1 ],
           ])

b=np.array([80,40,60,-70,-40,-70])

opt.linprog(c,A,b)

C:\Users\tony_\AppData\Local\Temp\ipykernel_13460\2429909114.py:16: OptimizeWarning: Solving system with option 'cholesky':True failed. It is normal for this to happen occasionally, especially as the solution is approached. However, if you see this frequently, consider setting option 'cholesky' to False.
  opt.linprog(c,A,b)
C:\Users\tony_\AppData\Local\Temp\ipykernel_13460\2429909114.py:16: OptimizeWarning: Solving system with option 'sym_pos':True failed. It is normal for this to happen occasionally, especially as the solution is approached. However, if you see this frequently, consider setting option 'sym_pos' to False.
  opt.linprog(c,A,b)


     con: array([], dtype=float64)
     fun: 539.9999999945517
 message: 'Optimization terminated successfully.'
     nit: 6
   slack: array([ 8.83190410e-10,  4.27569091e-10,  6.52448762e-10, -7.69162511e-10,
       -4.24861923e-10, -7.69190933e-10])
  status: 0
 success: True
       x: array([3.00000000e+01, 4.00000000e+01, 1.00000000e+01, 4.00000000e+01,
       4.67449811e-11, 6.07990748e-12, 2.20972261e-11, 2.37736910e-11,
       6.00000000e+01])

Para que el ***costo mínimo*** de abastecer a esas ciudades ***sea de 540 pesos*** las centrales deben abastecer:
- De la planta 1 a Guadalajara 30 millones de Kw
- De la planta 1 a León, Gto 40 millones de Kw
- De la planta 1 a Morelia 10 millones de Kw
- La planta 2 debe enviar toda su producción (40 millones de Kw) a Guadalajara
- La planta 3 debe enviar toda su producción (60 millones de kw) a Morelia